staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/FileSystemDao.java

```diff
@@ -282,6 +290,45 @@ public void cleanUpStaleInprogressBatches()
         }
     }
 
+    @Override
+    public BatchStatusResponse getBatchStatus(final TenantId tenantId, final BatchId batchId)
+        throws BatchNotFoundException, InterruptedException
+    {
+        final BatchStatusResponse responseBean = new BatchStatusResponse();
+        responseBean.setBatchId(batchId.getValue());
+        final BatchStatus batchStatus = new BatchStatus();
+        final InProgress inProgress = new InProgress();
+        final boolean batchInCompletedState = batchPathProvider.getPathForBatch(tenantId, batchId).toFile().exists();
+        batchStatus.setBatchComplete(batchInCompletedState);
+        final List<InProgressMetrics> inProgressMetricsList = new ArrayList<>();
+        //get threadIDs for requested batchId from filesystem
+        final List<String> listOfInProgressBatches = batchPathProvider.getListOfInProgressThreadFromFileSystem(tenantId, batchId);
+        if (!batchInCompletedState && listOfInProgressBatches.isEmpty()) {
+            throw new BatchNotFoundException(batchId.getValue());
+        }
+        for (final String batchThread : listOfInProgressBatches) {
+            final InProgressMetrics inProgressMetrics = new InProgressMetrics();
+            if (!batchThread.contains(ServiceIdentifier.getServiceId())) {
+                batchProgressTracker.putAll(batchPathProvider.monitorBatchProgressOfDifferentServices(batchThread));
+            }
+            final Tracker tracker = batchProgressTracker.get(batchThread);
```

Comment:
The code is effectively looking up by thread.  But how do you know that thread hasn't moved on to a different batch?  The thread could potentially have even moved on to dealing with a different tenant between the time you retrieved these ids and the time you are attempting to retrieve these progress reports.  So this is not guaranteed to be giving back progress reports for the correct batch.

Reply:
**Resolved in this commit:**
[590f331c0db34ae520d5a2d8a635c62a07f8bd09] Code review: Extensive refactoring

The map has been updated so that it is additionally keyed by tenant and batch rather than just by thread id.